### ISSUE TITLE
Add utilities to improve QuantityType handling

### ISSUE THREAD

**Issue by rkoshak** [openhab/openhab-js#102]
QuantityTypes are somewhat of a pain to work with. I think the current implementation in openhab-js leaves much to be desired.

My understanding is that currently the library goes out of it's way to throw away the units for QuantityTypes and pretty much forces the users who actually want to use units in their calculations to revert to messing with the rawState and importing and creating QuantityTypes themselves. 

This issue is to discuss and track ideas for ways to make it easier for end users to handle QuantityTypes.

Unfortunately, based on my research and experiments I've determined the following:

- JavaScript does not support operator overloading so we can't patch something in to make +, -, et. al. handle this.
- Try as I might, I could not monkey patch a Java class.

So we are looking at adding `add()`, `subtract()`, et. al. helper functions *somewhere*.

I've already done some experimentation using this approach and I think it could work. What I propose is adding all the operations as helper functions (to mirror the Java QuantityType I'd just create a compareTo instead of equals, lt, gt, et. al. functions). Each of these functions accept two arguments and follow these rules:

- if both are QT, make sure they are compatible units and if so, do the operation, throw an exception otherwise
- if one is any sort of number and the other is QT, convert the number to a QT of the same units and do the operation
- if both are any sort of number, do the operation as normal
- if one is a string representing a number with units (e.g. '65 °F'), convert that to a QT and then follow the rules above.

In experimenting I've basically already implemented everything but that last rule in a prototype and it all works as expected (though there are some gotchas with multiply and divide which I may need to think about, namely is it even reasonable to convert the non-QT operand to a QT?).

Before I go and create a new PR (given the trouble I've had already I want to wait until my outstanding PRs are merged before creating new ones) I want to know:

1. Does this sound like a good idea? It's probably going to be less useful to people who are writing rules in files or who are very well aware of the types of their Items but I think it's going to be critical for Blockly users when Blockly moves to JS Scripting.
2. Where to put it? utils.js? Item?

## Your Environment

* Version used: (e.g., openHAB and **JS Scripting add-on version**)
* Environment name and version (e.g. Chrome 76, Java 8, Node.js 12.9, ...):
* Operating System and version (desktop or mobile, Windows 10, Raspbian Buster, ...):

**Comment by jpg0** [comment]
I'm posting now as I've only just been learning about QuantityTypes...

It seems that your proposed implementation would copy all the existing types from the Java side. Would it be worth taking a look at using JS proxies instead, so that you can wrap the Java object into a JS one and simply pass through method calls, rather than having to reimplement each one?

**Comment by digitaldan** [comment]
@rkoshak can you give a few examples of how you are using Quanity types (or how you would like to use them)?  I have to admit, i don't actually pay attention to them much and want to make sure i see it from your point of view.

**Comment by rkoshak** [comment]
Circling back around to thing since the time utils PR was merged.

> It seems that your proposed implementation would copy all the existing types from the Java side. 

I not sure I addressed this but no, I'm not actually proposing copying the all the Java types. I think I'd only need to import QuantityType. 

So an `add(a, b)` utility function would do something like the following:

```
const QT = Java.type('org.openhab.core.library.types.QuantityType');

const add = function(a, b) {
    const aIsQT = a instanceof QT;
    const bIsQT = b instanceof QT;

    // Neither are QT
    if(!aIsQT && !bIsQT){
        // Maybe can do more here to test that a and b can be added and throw a meaningful error if not
        return a + b;
    }

    // Both are QT
    if(aIsQT && bIsQT) {
        return a.add(b);
    }

    // a is QT
    if(aIsQT) {
        // Probably need to do some further tests here and perhaps a parse to coerce b into a Number
        return a.add(new QT(b, a.getUnit()));
    }

    // b is QT
    // Probably need to do some further tests here and perhaps a parse to coerce b into a Number
    return b.add(new QT(a, b.getUnit());
}
```

There would be a similar implementation for all of the operations. I'm sure something clever could be done to avoid a bunch of duplicated code. The above is for illustrative purposes.

So you see, I'm not looking to import anything more than QuantityType. The function should work whether you have zero, one, or both operands being of type QuantityType. We assume that if only one operand is a QuantityType, the second operand is of the same units, which is the most common use case I've seen on the forum.

I'm not sure creating a proxy Object does anything for us here though. QuantityType itself doesn't handle the ability to do math unless both operands are of QuantityTypes. So we will have to implement something to make it so it works with non-QuantityTypes too. Whether it makes sense to do that implementation in a proxy Object I can go either way. My only concern is that we reimpose the need for the end users to have to do it one way if they have QuantityTypes and another way if they do not. That's the root problem I'm trying to solve.
